### PR TITLE
Psykal OpenMP taskloop implementation of NemoLite2D

### PR DESCRIPTION
**Pull request by LonelyCat124**
I have implemented a script using the `OMPTaskloopTrans` and `OMPTaskwaitDirective`.

I think this correctly captures all of the dependencies, while minimising the number of statements required to do this. This is done by lines 31-62 of the code, and lines 68-70.

What this does, is use `forward_dependence` to find the next dependency for each loop, and find the index of this dependency in the schedule. It then loops through these dependencies. If there are other dependencies between X and its dependent which are covered by the creation of such a dependence, it is deleted (line 61-62). If another dependence will be created sooner, this dependence is removed. 

I'm sure there are some cases which this doesn't entirely minimise, so if anyone can think of those examples (or ways to implement tests of these?). I think this still handles a case such as:
```
[ 5, 4, 3, None, None, None, None, ...]
```
which would turn into:
```
[ None, 4, 3, None, None, None, None, ...]
[ None, None, 3, None, None, None, ...]
```

This requires https://github.com/stfc/PSyclone/pull/1368 to work right now.

The dependence analysis here is a prototype that will be used to build the `OMPTaskwaitTransformation` at a later stage. I am going to add comments to explain the process now too.

### CODE REVIEW

**Comment by LonelyCat124** [comment]
I have an implementation now supporting distributed memory as well which I'll push once the VPN is working for me. I've not tested it with an up-to-date compiler, I will attempt on glados also once the VPN is working for me again.

**Comment by LonelyCat124** [comment]
Ok, this builds on Scafellpike with OpenMPI4 & gcc 9 so I'm happy the code produced is valid. I'll check correctness next.

**Comment by LonelyCat124** [comment]
Non-mpi version:
```
ua checksum =   0.52670513E+02
va checksum =   0.57100321E+03
```

psykal_omp manual version:
```
ua checksum =   0.52670513E+02
va checksum =   0.57100321E+03
```

MPI task version:
```
ua checksum =   0.52670513E+02
va checksum =   0.57100321E+03
```

Looks like its correct! Think this is ready for review.

**Comment by LonelyCat124** [comment]
Ok, I think this now works using PSyclone master branch. My only issue is that (on my laptop) the MPI build fails with an ICE in mpifort. I will test it on glados when my VPN is behaving properly.

@arporter  @sergisiso do either of you have time to have a quick look over this and check the script seems appropriate? The non-MPI version is pretty simple, the MPI one has to do a lot of checking and I don't know if there's a better way to do it.

**Comment by arporter** [comment]
Hi @LonelyCat124, I've had a look and I think it's OK. I think it's the same problem we tackle in the OpenACC script where we want to find the biggest region that we can put inside an ACC Kernels block, see https://github.com/stfc/PSyclone/blob/b1d9c0107057af563a782aeb8f8becbe9906627d/examples/nemo/scripts/kernels_trans.py#L456-L499 I'm not sure what we have there is much better than what you have though.

We should probably provide some sort of generalised support for implementing this pattern. A user could provide a function that determines whether it's OK to include a given node in the region and get back a bunch of regions.

**Comment by LonelyCat124** [comment]
Yeah it looks pretty similar.

Shall I make an issue in PSyclone for that functionality?

**Comment by arporter** [comment]
> 

Yes please!

**Comment by arporter** [comment]
Hi @LonelyCat124, presumably this now needs a full review? (Well, as 'full' as we do for PSycloneBench.)

**Comment by LonelyCat124** [comment]
I think its ready now yes!

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        # Find all of the groupings of taskloop and taskwait directives. Each of these
+        # groups needs its own parallel+single regions.
+        next_start = 0
+        next_end = 0
+        for index in range(len(schedule.children)):
+            child = schedule.children[index]
+            if isinstance(child, OMPTaskloopDirective) or isinstance(child, OMPTaskwaitDirective):
+                next_end = next_end + 1
+            elif not isinstance(child, OMPDirective):
+                if next_start == index:
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:51]: Please could you add some comments here.

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                    next_end = index + 1
+                    next_start = index + 1
+                else:
+                    sets.append((next_start, next_end))
+                    next_end = index + 1
+                    next_start = index+1
+            else:
+                next_end = next_end + 1
+        if next_start <= index:
+            sets.append((next_start, index+1))
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:61]: More comments please.

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        sets.reverse()
+        for next_set in sets:
+            single_trans.apply(schedule[next_set[0]:next_set[1]])
+            parallel_trans.apply(schedule[next_set[0]])
+        for child in schedule.children:
+            if isinstance(child, OMPParallelDirective):
+                wait_trans.apply(child)
+    else:
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:69]: I think I would put this branch first and then `return` from it. That way, the bulk of the code above here (which deals with DM) could be less indented.

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good. Thanks for also fixing  the various scripts which were out of date (we really need to run the PSyclone transformations as part of the CI but I think I already have an issue for that).
Main request is for more comments in the transformation script to help me understand what's going on.

File: benchmarks/nemo/nemolite2d/psykal/Makefile_gen [hunk 4]
```diff
@@ -93,7 +93,7 @@ clean:
 
 # OpenCL Execution helpers
 run_opencl_jit:
-	FORTCL_KERNELS_FILE=allkernels.cl ./nemolite2d.exe
+	 DL_ESM_ALIGNMENT=64 FORTCL_KERNELS_FILE=allkernels.cl ./nemolite2d.exe
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply benchmarks/nemo/nemolite2d/psykal/Makefile_gen:96]: I will chip in here. Technically the OpenCL global_size (which is controlled by DL_ESM_ALIGNMENT) must be divisible by the work_group_size (which is set to 64 in the psylcone script) in OpenCL implementations < 2.0. But from OpenCL 2.0 the standard says its not necessary anymore. In practice it is safe to always use it as I have seen most implementations are lacking in this aspect.

I would keep the change, but this number must be kept in sync with the ../psyclone_scprits/ocl_transform.py TILING parameter. Maybe you could rename TILING to WORK_GROUP_SIZE in that script and insert the consideration above and add the warning to update DL_ESM_ALIGNMENT in the comment before the parameter.

[the remark this answers is not in the document]
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/Makefile_gen:96]: I don't remember changing this so I'm not sure where its come from...

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+''' Python script intended to be passed to PSyclone's generate()
+function via the -s option. It applies OpenMP to every loop and
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:2]: Done

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+inlines all kernels in the schedule.'''
+
+from psyclone.psyGen import TransInfo
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:5]: Removed

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+from psyclone.psyir.nodes import Loop
+from psyclone.configuration import Config
+from psyclone.transformations import OMPParallelTrans, OMPSingleTrans, \
+                                     OMPTaskloopTrans, KernelModuleInlineTrans
+from psyclone.psyir.transformations import OMPTaskwaitTrans
+from psyclone.psyir.nodes import OMPTaskloopDirective, OMPTaskwaitDirective, \
+                                 OMPDirective, OMPParallelDirective
+from psyclone.psyGen import HaloExchange
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:13]: Removed

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def trans(psy):
+    '''Transformation entry point'''
+    config = Config.get()
+
+    schedule = psy.invokes.get('invoke_0').schedule
+
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:21]: Done

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+
+    loop_trans = OMPTaskloopTrans(grainsize=32, nogroup=True)
+    wait_trans = OMPTaskwaitTrans()
+
+    module_inline_trans = KernelModuleInlineTrans()
+
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:28]: Done

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    parallel_trans = OMPParallelTrans()
+    sets = []
+    if config.distributed_memory:
+        # Find all of the groupings of taskloop and taskwait directives. Each of these
+        # groups needs its own parallel+single regions.
+        next_start = 0
+        next_end = 0
+        for index in range(len(schedule.children)):
+            child = schedule.children[index]
+            if isinstance(child, OMPTaskloopDirective) or isinstance(child, OMPTaskwaitDirective):
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:48]: Fixed

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+    single_trans = OMPSingleTrans()
+    parallel_trans = OMPParallelTrans()
+    sets = []
+    if config.distributed_memory:
+        # Find all of the groupings of taskloop and taskwait directives. Each of these
+        # groups needs its own parallel+single regions.
+        next_start = 0
+        next_end = 0
+        for index in range(len(schedule.children)):
```
[the remark this answers is not in the document]
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:46]: Done

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py (hunk 3 above)
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:69]: Done

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py (hunk 2 above)
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:61]: Done

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py (hunk 1 above)
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:51]: Done

**Comment by LonelyCat124** [comment]
@arporter I think I fixed the review changes, but I've either broken this Makefile or I've broken my PSyclone installation. Can you double check this still builds for you?

File: benchmarks/nemo/nemolite2d/psykal/Makefile_gen (hunk 4 above)
[the remark this answers is not in the document]
Comment (arporter) [reply benchmarks/nemo/nemolite2d/psykal/Makefile_gen:96]: Thanks for the suggestions Sergi and for implementing them Aidan.

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py [hunk 12]
```diff
@@ -38,35 +33,50 @@ def trans(psy):
...
+            # grouping of OpenMP directives then we stop, and add it to
+            # the set of groupings. Otherwise we just skip over this
+            # node.
+            if next_start == idx:
+                next_end = idx + 1
+                next_start = idx + 1
+            else:
+                sets.append((next_start, next_end))
+                next_end = idx + 1
+                next_start = idx+1
```
Comment (arporter) [inline benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py:63]: Spaces around operator please.

**Review by arporter — CHANGES_REQUESTED** [review]
Looking good now - thanks for the comments.
Just an outstanding pylint issue to fix (use of enumerate) and one tiny formatting thing.
The nemolite2d_ompt target still compiles and runs fine for me.

File: benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py (hunk 12 above)
Comment (LonelyCat124) [reply benchmarks/nemo/nemolite2d/psykal/psyclone_scripts/omp_task_transform.py]: Fixed

**Comment by LonelyCat124** [comment]
I think I fixed the pylint issue, and had to add an extra declaration of `idx` so it was visible outside the loop.

**Review by arporter — APPROVED** [review]
All good now. I can still compile and run the nemolite2d code with and without MPI.
The only possible issue is that I don't seem to be able to make use of more than 4 cores when running with MPI plus threading. Perhaps it's a binding problem.

**Pull request merged**